File: modules/util/jvm/src/main/scala/coursier/parse/CredentialsParser.scala
package coursier.parse

import coursier.credentials.DirectCredentials
import coursier.util.Traverse._
import coursier.util.ValidationNel

object CredentialsParser {

  private val pattern = """([a-zA-Z0-9.-]*)(\(.*\))?[ ]+([^ :][^:]*):(.*)""".r.pattern

  def parse(s: String): Either[String, DirectCredentials] = {
    val m = pattern.matcher(s)
    if (m.matches()) {
      val cred = DirectCredentials(m.group(1), m.group(3), m.group(4))
          .withRealm(Option(m.group(2)).map(_.stripPrefix("(").stripSuffix(")")))
      Right(cred)
    } else
      Left("Malformed credentials") // FIXME More precise error message?
  }

  def parseSeq(input: String): ValidationNel[String, Seq[DirectCredentials]] =
    Predef.augmentString(input)
      .lines
      .map(_.dropWhile(_.isSpaceChar)) // not trimming chars on the right (password)
      .filter(_.nonEmpty)
      .toVector
      .validationNelTraverse { s =>
        val e = parse(s)
        ValidationNel.fromEither(e)
      }

}

File: modules/bootstrap-launcher/src/main/java/coursier/bootstrap/launcher/credentials/CredentialsParser.java
package coursier.bootstrap.launcher.credentials;

import java.util.Arrays;
import java.util.List;
import java.util.Optional;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

/**
 * Java copy of coursier.parse.CredentialsParser
 */
public class CredentialsParser {

  private static final Pattern CREDENTIALS_PARSER_PATTERN = Pattern
      .compile("([a-zA-Z0-9.-]*)(\\(.*\\))?[ ]+([^ :][^:]*):(.*)");

  public static final Optional<DirectCredentials> parse(String s) {
    final Matcher m = CREDENTIALS_PARSER_PATTERN.matcher(s);
    if (m.matches()) {
      return Optional.of(new DirectCredentials(m.group(1), m.group(3), m.group(4))
        .withRealmOpt(Optional.ofNullable(m.group(2)).map(r -> r.replaceFirst("^\\(", "").replaceFirst("\\)$", "")).orElseGet(() -> null)));
    } else {
      return Optional.empty();
    }
  }

  public static final List<DirectCredentials> parseList(String input) {
    return Arrays.stream(input.split("\\r?\\n")) // Java 11: input.lines()
      .map(s -> s.replaceFirst("^\\s+", "")) // not trimming chars on the right (password)
      .filter(s -> !s.isEmpty())
      .map(CredentialsParser::parse)
      .filter(o -> o.isPresent())
      .map(o -> o.get())
      .collect(Collectors.toList());
  }

}
